file.cc

```diff
@@ -51,179 +33,95 @@ bool tryPossibleZipName( std::string const & name, std::string & copyTo )
-                openMode |= QIODevice::ReadWrite;
-                break;
-      case 'a': openMode |= QIODevice::Append;
-                break;
-      case 'b': openMode &= ~QIODevice::Text;
-                break;
-      default:  break;
+  while ( *pch ) {
+    switch ( *pch ) {
+      case 'r':
+        openMode |= QIODevice::ReadOnly;
+        break;
+      case 'w':
+        openMode |= QIODevice::WriteOnly;
+        break;
+      case '+':
+        openMode &= ~( QIODevice::ReadOnly | QIODevice::WriteOnly );
+        openMode |= QIODevice::ReadWrite;
+        break;
+      case 'a':
+        openMode |= QIODevice::Append;
+        break;
+      case 'b':
+        openMode &= ~QIODevice::Text;
+        break;
+      default:
+        break;
     }
     ++pch;
   }
 
-  f.setFileName( filename );
-
   if ( !f.open( openMode ) )
-    throw exCantOpen( std::string( filename ) + ": " + f.errorString().toUtf8().data() );
+    throw exCantOpen( f.fileName().toStdString() + ": " + f.errorString().toUtf8().data() );
 }
 
-Class::Class( char const * filename, char const * mode ) :
-  writeBuffer( 0 )
+Class::Class( std::string_view filename, char const * mode )
 {
-  open( filename, mode );
+  f.setFileName( QString::fromStdString( filename.data() ) );
+  open( mode );
 }
 
-Class::Class( std::string const & filename, char const * mode )
-  : writeBuffer( 0 )
+void Class::read( void * buf, qint64 size )
 {
-  open( filename.c_str(), mode );
-}
-
-void Class::read( void * buf, qint64 size ) 
-{
-  if ( !size )
-    return;
-
-  if ( writeBuffer )
-    flushWriteBuffer();
-
-  qint64 result = f.read( reinterpret_cast<char *>( buf ), size );
+  qint64 result = f.read( static_cast< char * >( buf ), size );
 
   if ( result != size )
     throw exReadError();
 }
 
-size_t Class::readRecords( void * buf, qint64 size, size_t count ) 
+size_t Class::readRecords( void * buf, qint64 size, qint64 count )
 {
-  if ( writeBuffer )
-    flushWriteBuffer();
-
-  qint64 result = f.read( reinterpret_cast<char *>( buf ), size * count );
+  qint64 result = f.read( static_cast< char * >( buf ), size * count );
   return result < 0 ? result : result / size;
 }
 
-void Class::write( void const * buf, qint64 size ) 
+void Class::write( void const * buf, qint64 size )
 {
-  if ( !size )
-    return;
-
-  if ( size >= WriteBufferSize )
-  {
-    // If the write is large, there's not much point in buffering
-    flushWriteBuffer();
-
-    qint64 result = f.write( reinterpret_cast<char const *>( buf ), size );
-
-    if ( result != size )
-      throw exWriteError();
-
-    return;
-  }
-
-  if ( !writeBuffer )
-  {
-    // Allocate the writing buffer since we don't have any yet
-    writeBuffer = new char[ WriteBufferSize ];
-    if( !writeBuffer )
-      throw exAllocation();
-    writeBufferLeft = WriteBufferSize;
-  }
-
-  size_t toAdd = size < writeBufferLeft ? size : writeBufferLeft;
+  if ( size <= 0 ) {
+    throw exWriteError();
```

Comment:
I think 

size<0 can throw exception
size=0 should return directly.

Reply:
Fixed